ci(check-ci-skip): fix commitMessagesMetadata.forEach is not a function

zondervancalvez opened the pull request:
## Commit to be reviewed

ci(check-ci-skip): fix commitMessagesMetadata.forEach is not a function
```
Primary Changes
----------------
1. Changed the method in getting the commit
message from GitHub API to shell command to avoid
the rate limits in calling the API.
2. Same goes for the author of commit message,
we use shell command to fetch the username.
```
Fixes #3614

**Pull Request Requirements**
- [ ] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficient and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when and, why.
- [ ] Have git sign off at the end of commit message to avoid being marked red. You can add `-s` flag when using `git commit` command. You may refer to this [link](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information.
- [ ] Follow the Commit Linting specification. You may refer to this [link](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) for more information. 

**Character Limit**
- [ ] Pull Request Title and Commit Subject must not exceed 72 characters (including spaces and special characters).
- [ ] Commit Message per line must not exceed 80 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

jagpreetsinghsasan reviewed (changes requested):
@zondervancalvez I don't think you have checked the code difference when creating this PR. Its totally wrong as you have already commented all the input arguments and hard-coded them with constant fields.
Also the check you have added should output a result which would be self explanatory. Saying `Commit message data is empty or null` doesn't help the creator of a PR. The right message should be `Fetching data from github API failed. This is most likely due to an intermittent issue from Github end or the api being rate limited (one hour cooldown). Please retry running the failed check after an hour.`
Also, it would be much better if we fine tune the error handling so that we can clearly say if it is an intermittent issue or rate limit issue.
Although rate limit errors are less likely to occur as the limit for the hosted runners is 1000 per hour

Also, even after mentioning it multiple times, yet we are at the same point where I have to specifically mention that `## Commit` and `## **Commit**` are rendered same.

petermetz commented on tools/ci-skip-for-maintainers.js:
```diff
@@ -33,10 +33,16 @@ const main = async () => {
+      commitMessageList.push(commitMessageMetadata["commit"]["message"]);
+    });
+  } else {
+    console.error(
```
@zondervancalvez Instead of just logging on stderr we should crash. Otherwise there will be more unintended side-effects to the error.
The bottom line is that if we canot fetch the commit message metadata, then we cannot complete the check at all.

@zondervancalvez Is it possible to refactor this code to use the local git installation and list the commits that way so that we are not hitting the GitHub API at all for this?

petermetz reviewed (changes requested):
@zondervancalvez Is this diff going to also fix the issue that if the check script crashes, the CI jobs get skipped by default?

petermetz reviewed (changes requested):
@zondervancalvez Do these changes solve the problem that if the ci-ckip-for-maintainers.js code crashes, it implicitly skips all the test checks?

jagpreetsinghsasan commented on tools/ci-skip-for-maintainers.js:
```diff
@@ -15,41 +15,29 @@ const MAINTAINERS_REGEX = new RegExp(
+    commitMessage = execSync("git log -1 --pretty=%B").toString().trim();
+    console.log("Latest commit message:\n", commitMessage);
+  } catch (error) {
+    console.error("Error fetching commit message:\n", error.message);
```
@petermetz I think the `git log..` will not fail at all for an already initialized repository and git binary exists as a pre-requisite (or else one can't even commit the changes/pull their forked repo). So the `try...catch` block is irrelevant here. @zondervancalvez can you please remove this block itself?

zondervancalvez commented:
> @zondervancalvez Could you please address the questions above?
Hi @petermetz please see the latest commit. Thanks

petermetz commented:
> > @zondervancalvez Could you please address the questions above?
> > Hi @petermetz please see the latest commit. Thanks

@zondervancalvez I'm sorry, I can't find the answer to my question from earlier regarding the crashes. Could you please be more specific about which part of the commit message is answering that question?

zondervancalvez commented on tools/ci-skip-for-maintainers.js:
```diff
@@ -12,90 +12,82 @@ const MAINTAINERS_REGEX = new RegExp(
-  //const committerLogin = "<username>";
+// Function to get the latest commit message
+const getLatestCommitMessage = () => {
+  const commitMsgBuffer = execSync("git log -1 --pretty=%B");
```
@petermetz to answer your question, this `git log` shell command is the alternative we used for fetching commit messages instead of calling github API to avoid limits and crashing

petermetz replied:
@zondervancalvez I'll post a screenshot here with the question(s). Maybe you posted the answers somewhere else and I just couldn't find it? Pull request comments/reviews get confusing quick.

![image](https://github.com/user-attachments/assets/7bb0d02f-74e1-4cb9-85af-af340e00f9aa)

zondervancalvez replied:
@petermetz I added `if: always()` on the jobs that requires `check-ci-skip` job. This will bypass check-ci-skip job if the job failed/crashes.

petermetz replied:
@zondervancalvez OK, but would that then mean that the skip CI marker does not work anymore?

What I mean is that the .js script currently exists with a job failure when it is trying to indicate that the CI should be skipped. But with `if: always()` the jobs will ignore that. If I'm understanding it correctly.

So I'm thinking you need to refactor the .js script to use a different signaling mechanism where it succeeds as long as there was no crash in it, and it places the information about the CI being skipped  or not into some other context variable  (which must not be writeable by pull request code because then people could just hack it by writing to this place where we store the information about the CI being skipped or not)

```js
if (isMaintainer) {
  console.log(
    "CI skipped as per request. Exit with an error to PAUSE dependent workflows.",
  );
  process.exit(1);
} else {
  process.exit(0);
}
```

petermetz reviewed (changes requested):
@zondervancalvez Please see https://github.com/hyperledger-cacti/cacti/pull/3618#discussion_r1884788850

petermetz reviewed (changes requested):
@zondervancalvez Sorry for just noticing this, but we also have a security issue with parsing the git logs directly (my fault, I didn't realize this when I was suggesting the workaround). The problem is that the author/comitter fields in the git log can be spoofed because we don't have commit signature validation in place. So for now, please evaluate if it would be possible to get the username of the pull request author extracted (without the GitHub API calls which would then break the job again). My idea is to use the process.env.GITHUB_ACTOR environment variable, but please do verify that this actually contains the information we want it to in the context of a pull request.

Note to self for later: We could introduce an allow-list of public keys in the GitHub org's secrets and then check the commit signature against the public key's from the allow list. That way we would have much fewer attack vectors. We can refactor this as a second version of the job, for now just please look into the point I made in my previous paragraph.

zondervancalvez commented:
Hi @petermetz, I have addressed your concern. I’ve replaced the previous approach with process.env.GITHUB_ACTOR, which directly provides the username of the person who initiated the workflow. Let me know if you have any further suggestions.

jagpreetsinghsasan reviewed (commented):
LGTM given all is properly tested this time @zondervancalvez

petermetz reviewed (commented):
Closing this for now until we can determine next steps.
